spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/quantized/SpatialConvolution.scala

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn.quantized
+
+import com.intel.analytics.bigdl.bigquant.BigQuant
+import com.intel.analytics.bigdl.nn.ErrorInfo
+import com.intel.analytics.bigdl.nn.abstractnn.{DataFormat, Initializable, TensorModule}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor._
+import com.intel.analytics.bigdl.utils.serializer.{DataConverter, ModuleData}
+import com.intel.analytics.bigdl.utils.{T, Table}
+import java.io.{IOException, ObjectInputStream, ObjectOutputStream}
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe
+import serialization.Bigdl.{AttrValue, BigDLModule}
+
+@SerialVersionUID(- 8008252944905538960L)
+private[bigdl] class SpatialConvolution[T: ClassTag](
+  val nInputPlane: Int, // The number of expected input planes in the image given into forward()
+  val nOutputPlane: Int, // The number of output planes the convolution layer will produce.
+  val kernelW: Int, // The kernel width of the convolution
+  val kernelH: Int, // The kernel height of the convolution
+  val strideW: Int = 1, // The step of the convolution in the width dimension.
+  val strideH: Int = 1, // The step of the convolution in the height dimension
+  val padW: Int = 0, // The additional zeros added per width to the input planes.
+  val padH: Int = 0, // The additional zeros added per height to the input planes.
+  val nGroup: Int = 1, // Kernel group number
+  val format: DataFormat = DataFormat.NCHW
+)(implicit ev: TensorNumeric[T]) extends QuantizedModule[T](nOutputPlane) with Initializable {
+
+  require(nInputPlane % nGroup == 0, "Number of input channels should be multiples of group.")
+  require(nOutputPlane % nGroup == 0, "Number of output channels should be multiples of group.")
+
+  @transient var weight: Array[Tensor[T]] = null
+  var data: QuantizedTensor[T] = QuantizedTensor[T]()
+  val bias: Tensor[T] = Tensor[T](nOutputPlane)
+
+  val quantFormat = if (format == DataFormat.NCHW) {
+    BigQuant.NCHW
+  } else {
+    BigQuant.NHWC
```

Comment:
Have you tested NHWC?

Reply:
Yes. There's a test case for data format.